### ISSUE TITLE
sf_coordinates?

### ISSUE THREAD

**Issue by Robinlovelace** [dcooley/sfheaders#52]
My attempt at creating such a function resulted in a slowdown. Such a simple extension of `sf_to_df()` could be useful and perhaps even faster I imagine...

``` r
# Aim: test sfheaders vs sf for getting coordinates
p = stplanr::cents_sf
df1 = sf::st_coordinates(p)
class(df1)
#> [1] "matrix"
df1
#>           X        Y
#> 1 -1.546463 53.80952
#> 2 -1.511861 53.81161
#> 3 -1.524205 53.80410
#> 4 -1.550806 53.82442
#> 5 -1.535617 53.82847
#> 6 -1.516734 53.82887
#> 7 -1.519318 53.81580
#> 8 -1.530712 53.81756
df2 = sfheaders::sf_to_df(p)
class(df2)
#> [1] "data.frame"
df2
#>   sfg_id point_id         x        y
#> 1      1        1 -1.546463 53.80952
#> 2      2        2 -1.511861 53.81161
#> 3      3        3 -1.524205 53.80410
#> 4      4        4 -1.550806 53.82442
#> 5      5        5 -1.535617 53.82847
#> 6      6        6 -1.516734 53.82887
#> 7      7        7 -1.519318 53.81580
#> 8      8        8 -1.530712 53.81756
sf_coordinates = function(x) {
  coords = sfheaders::sf_to_df(x)[c("x", "y")]
  names(coords) = c("X", "Y")
  as.matrix(coords)
} 
df3 = sf_coordinates(p)
df3
#>           X        Y
#> 1 -1.546463 53.80952
#> 2 -1.511861 53.81161
#> 3 -1.524205 53.80410
#> 4 -1.550806 53.82442
#> 5 -1.535617 53.82847
#> 6 -1.516734 53.82887
#> 7 -1.519318 53.81580
#> 8 -1.530712 53.81756
identical(df1, df3)
#> [1] TRUE

bench::mark(check = FALSE,
  sf::st_coordinates(p),
  sfheaders::sf_to_df(p),
  sf_coordinates(p)
)
#> # A tibble: 3 x 6
#>   expression                  min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>             <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 sf::st_coordinates(p)    37.5µs   40.9µs    21988.        0B     11.0
#> 2 sfheaders::sf_to_df(p)   17.8µs   19.9µs    48539.    2.49KB     19.4
#> 3 sf_coordinates(p)        66.7µs     72µs    12412.   38.63KB     12.6
```

<sup>Created on 2020-02-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by dcooley** [comment]
How do you see this `sf_coordinates()` function handling mixed geometries?

Take this example of an `sf_GEOMETRY` of a POINT and  LINESTRING (with ZM dimension)
```r
sf_point <- sfheaders::sf_point(1:2)
sf_point$id <- 1
sf_line <- sfheaders::sf_linestring(1:4)

library(sf)
sf <- rbind( sf_point, sf_line )
sf
# Simple feature collection with 2 features and 1 field
# geometry type:  GEOMETRY
# dimension:      XY
# bbox:           xmin: 1 ymin: 2 xmax: 1 ymax: 2
# z_range:        zmin: NA zmax: NA
# m_range:        mmin: NA mmax: NA
# epsg (SRID):    NA
# proj4string:    NA
#   id                geometry
# 1  1             POINT (1 2)
# 2  1 LINESTRING ZM (1 2 3 4)

sfheaders::sf_to_df( sf )
#   sfg_id linestring_id point_id x y  z  m
# 1      1            NA        1 1 2 NA NA
# 2      2             2       NA 1 2  3  4
```
should your requested `sfheaders::sf_coordinates()` behave in exactly the same way as `sf::st_coordinates()` (i.e, it doesn't work on this type of object)?

```r
sf::st_coordinates( sf )
# Error in st_coordinates.sfc(st_geometry(x)) : 
#   not implemented for objects of class sfc_GEOMETRY
```

Or do you think it's worth keeping the structure of the `data.frame` and have `NA`s in the various ID columns to indicate the geometry doesn't exist for that row?

**Comment by Robinlovelace** [comment]
> should your requested sfheaders::sf_coordinates() behave in exactly the same way as sf::st_coordinates() (i.e, it doesn't work on this type of object)?

Good question, I hadn't thought about that. First thought: it could be useful to have the coordinates of geometry collection objects, although I cannot think of any use case at the moment. If `sf` fails (I didn't know that!) it would not be a divergence.

> Or do you think it's worth keeping the structure of the data.frame and have NAs in the various ID columns to indicate the geometry doesn't exist for that row?

I was thinking that returning *only* the columns returned by `st_coodinates()` could be useful. Another e.g.:

``` r
sf::st_coordinates(spData::lnd[1:2, ])
#>                     X        Y L1 L2 L3
#>    [1,] -3.306791e-01 51.32901  1  1  1
#>    [2,] -3.305944e-01 51.32909  1  1  1
#>    [3,] -3.305061e-01 51.32915  1  1  1
#>    [4,] -3.304092e-01 51.32921  1  1  1
#>    [5,] -3.302921e-01 51.32927  1  1  1
#>    [6,] -3.299135e-01 51.32946  1  1  1
#>    [7,] -3.296336e-01 51.32966  1  1  1
#>    [8,] -3.294288e-01 51.32984  1  1  1
#>    [9,] -3.293209e-01 51.32997  1  1  1
#>   [10,] -3.292145e-01 51.33009  1  1  1
#>   [11,] -3.291163e-01 51.33026  1  1  1
#>   [12,] -3.291130e-01 51.33040  1  1  1
#>   [13,] -3.291712e-01 51.33073  1  1  1
#>   [14,] -3.291913e-01 51.33097  1  1  1
#>   [15,] -3.293029e-01 51.33153  1  1  1
#>   [16,] -3.293628e-01 51.33162  1  1  1
#>   [17,] -3.295546e-01 51.33187  1  1  1

```

<sup>Created on 2020-02-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by Robinlovelace** [comment]
The output of `sf_to_df()` is slightly different. I think it could be useful to have a function that has equivalent, or near-equivalent, output as that demonstrated above for `st_coodinates()` (although I confess I was thinking only of the much simpler case of point geometries): 

``` r
res = sfheaders::sf_to_df(spData::lnd)
head(res)
#>   sfg_id multipolygon_id polygon_id linestring_id          x        y
#> 1      1               1          1             1 -0.3306791 51.32901
#> 2      1               1          1             1 -0.3305944 51.32909
#> 3      1               1          1             1 -0.3305061 51.32915
#> 4      1               1          1             1 -0.3304092 51.32921
#> 5      1               1          1             1 -0.3302921 51.32927
#> 6      1               1          1             1 -0.3299135 51.32946
```

<sup>Created on 2020-02-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by Robinlovelace** [comment]
And I wonder if the resulting `sf_coodinates()` could usefully result in a PR into `sf`...

**Comment by dcooley** [comment]
proof-of-concept 

```r
library(mapdeck)

sf <- mapdeck::roads

microbenchmark::microbenchmark(
  sf = { res <- sf::st_coordinates( sf ) },
  sfh = { res <- sfheaders:::rcpp_sfc_coordinates( sf$geometry ) },
  times = 5
)

# Unit: milliseconds
# expr      min       lq     mean   median       uq      max neval
#   sf 82.31320 83.20740 88.23467 90.21316 90.52957 94.91003     5
#  sfh 28.08056 28.35004 33.27829 33.92859 37.15311 38.87913     5

head( res )
#      [,1] [,2]     [,3]      [,4]
# [1,]    1    1 145.0143 -37.83046
# [2,]    1    1 145.0143 -37.83057
# [3,]    1    1 145.0145 -37.83070
# [4,]    1    1 145.0160 -37.83148
# [5,]    1    1 145.0165 -37.83170
# [6,]    1    1 145.0168 -37.83175
```
even though this is quicker than `sf`, I haven't optimised it for returning the matrix, I've just used the `to_df` code and stripped the df attributes. Therefore I recon I can get this even quicker by limiting the goemtry objects to a single type (as per `sf::st_coordinates()`)

**Comment by Robinlovelace** [comment]
Looks promising to me! I will likely be an early user of this new function if you put it in, and am very happy to test.

**Comment by dcooley** [comment]
I've got a long flight today so I'll try and get it done :)

**Comment by dcooley** [comment]
Notes for myself on design:

- the `_to_df()` and `_coordinates()` code will share a lot of common features
- folder structure
  - `/df/`
  - `/matrix`
  - `/coordinates`
- put the common code into `coordinates.hpp`
- the `matrix.hpp` will only allow one geometry type (as per `st_coordinates`)
- the `df.hpp` will continue to work as-is, and will accept a mix of geometries

**Comment by dcooley** [comment]
Some thoughts as I fly over the pacific drinking whiskey...

- following `sf::st_coordinates()` exaclty seems like a backwards step, as it has an underlying assumptionns that the first `sfg` object inside the `sfc` list determines the dimension of all the other `sfg` objects. Whereas one of my design choices for `sfheaders` is to be completely general and not have various `if { } else { }` conditions (where I can help it)
- So I think it's better programatically to allow mixed geometries (as per [this comment](https://github.com/dcooley/sfheaders/issues/52#issuecomment-583891505) )

So I don't know how best to approach this...

I understand the desire to have a faster replacement to `sf::st_coordinates()`, but I think I fundamentally disagree with how it works. 

@Robinlovelace - can I go back to basics and ask **why** you need a matrix, and is it a big impost on you to use the result of `sf_to_df()` ?

**Comment by Robinlovelace** [comment]
Don't need a matrix but the process of selecting only the 2 (X, Y) columns from the whole result returned by `sf_to_df()` adds complexity and slows things down compared with the proposed `sf_coordinates()` as shown in the benchmark in the original post. If there are quicker ways to select the needed columns that could also help.

The use case is a package called `od` that I plan to open source soon. See here for context https://github.com/ropensci/stplanr/issues/367

**Comment by Robinlovelace** [comment]
More specific answer: here I create the data frame and subset it in one command which is computationally sub-optimal I think: https://github.com/Robinlovelace/od/blob/master/R/od-funs.R#L86

As you'll see from the readme, the package's functions are already faster than the equivalents in `stplanr` but see opportunities to make it faster still and suspect others would find an `sf_coordinates()` functions useful, especially if they just want to extract the lon/lats from point objects as I do in this case.

**Comment by SymbolixAU** [comment]
I've been investigating this a bit more and here are some findings

## TL:DR

- if you only have POINTs then use the internal code from `sf::st_coordinates()`
- for other types of geometries use `sfheaders::sf_to_df()` 

---

1. `sfheaders::sf_to_df()` is actually slower than `sf::st_coordinates()` for larger `sfc_POINT` objects

```r
library(sf)
library(sfheaders)

n <- 1e6
df <- data.frame(
  x = rnorm(n)
  , y = rnorm(n)
)

pts <- sfheaders::sf_point(
  obj = df
)

sf <- sf::st_coordinates( pts )
sfh <- sfheaders::sf_to_df( pts )

all.equal( unname( as.matrix( sfh[, c("x","y")] ) ), unname( sf ) )
# [1] TRUE

microbenchmark::microbenchmark(
  sf = { sf::st_coordinates( pts ) }
  , sfh = { sfheaders::sf_to_df( pts ) }
  , times = 5
)

# Unit: milliseconds
# expr        min         lq       mean     median         uq        max neval
#  sf   22.51863   28.74512   31.36629   31.86352   35.98059   37.72361     5
# sfh 1063.10201 1366.02118 1322.52039 1391.10467 1394.05941 1398.31468     5

```

2. `sfheaders::sf_to_df()` is faster for other type of geometries, and mixed geometries [e.g.](https://github.com/dcooley/sfheaders/issues/28)

3. IFF you are only interested in `sfc_POINT` objects, then you really can't do much better than `sf::st_coordinates()`

- the internal code for `sf:::st_coordinates.sfc()` uses `base::unlist`, which itself is a highly optimised `.Internal` function. [Here is an example](https://stackoverflow.com/a/30201460/5977215) of Kevin Ushey's attempt to improve on it, but you can see from the results it really isn't much better.

- if you really do want to optimise it, AND you are only working with POINTs, then you can strip out the geometry checks from `sf::st_coordinates()` and just use the code of interest

```r
sf_internal <- function(x) {
  matrix(unlist(x, use.names = FALSE), 
         nrow = length(x), byrow = TRUE, dimnames = list(1:length(x)))
}

microbenchmark::microbenchmark(
  sf = { sf::st_coordinates( pts ) }
  , sfi = { sf_internal( pts$geometry ) }
  , times = 5
)

# Unit: milliseconds
# expr      min       lq     mean   median       uq      max neval
#  sf 36.44416 37.55566 43.91430 38.08560 39.00972 68.47635     5
# sfi 21.30363 30.37249 29.00011 30.59714 30.93934 31.78794     5

```

4. If you have other geometry types then `sfheaders` gives you a speed boost

```r
ls <- mapdeck::roads

microbenchmark(
  sf = { coords <- sf::st_coordinates( ls ) },
  sfh = { df <- sfheaders::sf_to_df( ls )[, c("x","y")] },
  times = 5
)

# Unit: milliseconds
# expr      min       lq     mean   median       uq      max neval
#  sf 74.62374 74.94685 77.45860 76.41045 76.64927 84.66267     5
# sfh 23.88828 24.56210 32.09882 25.45537 27.34708 59.24127     5
```

**Comment by Robinlovelace** [comment]
Thanks for the updated benchmark @dcooley (btw do you prefer being tagged as that, Symbolix or neither? realise I've been tagging you lots of late, hope it's useful and not annoying!) but I still see the `sfheaders` way in my code being faster, with the difference being **larger** on a medium sized dataset with 10k rows (also heads-up @mem48 who advocated this based on the plausible hypothesis that benchmarks on tiny datasets are unreliable): https://github.com/Robinlovelace/od#benchmark-on-medium-sized-dataset

I'll try to replicate your benchmark with this dataset to see what's going on. The provision of code that does not need `sf` or `sfheaders` is greatly appreciated.

**Comment by Robinlovelace** [comment]
Update, it seems the slowness is in `stplanr` code not `sf` code as you indicate based on this preliminary benchmark:

```
# A tibble: 10 x 14
   expression   rows      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
   <bch:expr>  <dbl> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
 1 sf             10   1.03ms   1.09ms   629.       6.36KB    0       315     0   500.87ms
 2 sfh            10      1ms   1.09ms   401.       8.24KB    0       202     0   503.12ms
 3 sf            100   3.34ms   3.61ms   215.      56.44KB    2.15    100     1   465.64ms
 4 sfh           100   3.52ms   3.66ms   244.      66.91KB    0       123     0   503.33ms
 5 sf           1000  26.89ms  29.18ms    31.5    474.55KB    0        16     0   507.34ms
 6 sfh          1000  28.41ms  32.97ms    29.5    551.82KB    2.27     13     1   441.11ms
 7 sf          10000 308.27ms 308.27ms     3.24     4.94MB    3.24      1     1   308.27ms
 8 sfh         10000 324.26ms 324.26ms     3.08     5.67MB    3.08      1     1   324.26ms
 9 sf         100000    3.21s    3.21s     0.311    47.3MB    1.24      1     4      3.21s
10 sfh        100000    3.36s    3.36s     0.298   54.55MB    0.894     1     3      3.36s
# … with 4 more variables: result <list>, memory <list>, time <list>, gc <list>
```

**Comment by Robinlovelace** [comment]
Hypothesis: gc, which I guess means garbage collection, is slowing your code down (this is way above my head but the green and blue dots seem to be further to the right than the red ones).

**Comment by Robinlovelace** [comment]
Another benchmark including the `sf_internal()` function and without the 1 million row dataset as a reprex:

```` r

sf_internal = function(x) {
  matrix(
    unlist(x, use.names = FALSE),
    nrow = length(x),
    byrow = TRUE,
    dimnames = list(1:length(x))
  )
}
# ```

# ```{r}
n = 1e5
df = data.frame(x = rnorm(n),
                y = rnorm(n))

pts = sfheaders::sf_point(obj = df)

sf = sf::st_coordinates(pts)
sfh = sfheaders::sf_to_df(pts)
sfi = sf_internal(pts$geometry)
head(sf)
#>             X           Y
#> 1 -0.77170979  0.09697509
#> 2  0.28746300 -0.46204060
#> 3 -0.03034466 -0.31921256
#> 4 -0.89545981  1.30896491
#> 5  0.20658428 -1.18993778
#> 6 -0.37709756  0.59974606
head(sfh)
#>   sfg_id point_id           x           y
#> 1      1        1 -0.77170979  0.09697509
#> 2      2        2  0.28746300 -0.46204060
#> 3      3        3 -0.03034466 -0.31921256
#> 4      4        4 -0.89545981  1.30896491
#> 5      5        5  0.20658428 -1.18993778
#> 6      6        6 -0.37709756  0.59974606
head(sfi)
#>          [,1]        [,2]
#> 1 -0.77170979  0.09697509
#> 2  0.28746300 -0.46204060
#> 3 -0.03034466 -0.31921256
#> 4 -0.89545981  1.30896491
#> 5  0.20658428 -1.18993778
#> 6 -0.37709756  0.59974606

all.equal(unname(as.matrix(sfh[, c("x", "y")])), unname(sf))
#> [1] TRUE
all.equal(unname(sfi), unname(sf))
#> [1] TRUE

res = bench::press(
  rows = 10 ^ (1:5),
  bench::mark(
    check = FALSE,
    sf = sf::st_coordinates(pts[1:rows, ]),
    sfh = sfheaders::sf_to_df(pts[1:rows, ]),
    sfi = sf_internal(pts[1:rows, ])
  )
)
#> Running with:
#>     rows
#> 1     10
#> 2    100
#> 3   1000
#> 4  10000
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.
#> 5 100000
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.

res
#> # A tibble: 15 x 7
#>    expression   rows      min   median `itr/sec` mem_alloc `gc/sec`
#>    <bch:expr>  <dbl> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#>  1 sf             10 997.02µs   1.11ms   624.     445.11KB    14.0 
#>  2 sfh            10 965.26µs   1.05ms   717.       8.24KB    17.7 
#>  3 sfi            10 942.02µs   1.03ms   685.       3.94MB    13.2 
#>  4 sf            100   3.27ms   3.71ms   151.      56.44KB     9.14
#>  5 sfh           100   3.32ms   3.78ms   214.      66.91KB    19.1 
#>  6 sfi           100   3.21ms   3.39ms   229.      54.83KB    16.2 
#>  7 sf           1000  26.57ms  27.97ms    30.4    474.55KB    30.4 
#>  8 sfh          1000   27.5ms  32.35ms    27.8    551.82KB    32.5 
#>  9 sfi          1000  26.39ms   31.6ms    29.3    458.88KB    34.2 
#> 10 sf          10000 378.69ms 417.13ms     2.40     4.94MB     9.59
#> 11 sfh         10000 342.52ms 348.88ms     2.87     5.67MB     5.73
#> 12 sfi         10000 318.08ms 341.94ms     2.92     4.79MB     4.39
#> 13 sf         100000    5.06s    5.06s     0.198    47.3MB     2.97
#> 14 sfh        100000    3.83s    3.83s     0.261   54.55MB     4.71
#> 15 sfi        100000    3.74s    3.74s     0.267   45.77MB     4.01
ggplot2::autoplot(res)
#> Loading required namespace: tidyr
#> Warning in f(...): The default behavior of beeswarm has changed in version
#> 0.6.0. In versions <0.6.0, this plot would have been dodged on the y-axis. In
#> versions >=0.6.0, grouponX=FALSE must be explicitly set to group on y-axis.
#> Please set grouponX=TRUE/FALSE to avoid this warning and ensure proper axis
#> choice.
````

![](https://i.imgur.com/2gfzqhP.png)

<sup>Created on 2020-02-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This raises the question: does the `sfheaders` code slow down substantially after ~1m rows?

**Comment by SymbolixAU** [comment]
> btw do you prefer being tagged as that, Symbolix or neither

@dcooley is probably better for tagging specifically me in a conversation, but I will sometimes reply as Symbolix if I am at worked (as that's what I'll be logged-in as). And for what it's worth, the SymbolixAU account is run by myself and Elizabeth

> This raises the question: does the sfheaders code slow down substantially after ~1m rows?

That is a very good question, and one I [need to look in to](https://github.com/dcooley/sfheaders/issues/56)

**Comment by SymbolixAU** [comment]
while working on #56 I've got the performance of `POINT` to `data.frame` more respectable

```r

library(sf)
library(sfheaders)

n <- 1e6
df <- data.frame(
  x = rnorm(n)
  , y = rnorm(n)
)

pts <- sfheaders::sf_point(
  obj = df
)

sf <- sf::st_coordinates( pts$geometry )
sfh <- sfheaders::sfc_to_df( pts$geometry )

all.equal( unname( as.matrix( sfh[, c("x","y")] ) ), unname( sf ) )
# [1] TRUE

microbenchmark::microbenchmark(
  sf = { sf::st_coordinates( pts$geometry ) }
  , sfh = { sfheaders::sfc_to_df( pts$geometry ) }
  , times = 5
)

# Unit: milliseconds
# expr       min        lq      mean    median       uq       max neval
#   sf  33.60898  43.52285  49.69699  48.69359  60.5833  62.07623     5
#  sfh 124.11240 135.55903 134.95872 136.46028 138.0602 140.60164     5

microbenchmark::microbenchmark(
  sf = { sf::st_coordinates( pts ) }
  , sfh = { sfheaders::sf_to_df( pts ) }
  , times = 5
)

# Unit: milliseconds
# expr       min        lq      mean    median        uq      max neval
#   sf  22.66758  22.86338  33.97696  39.09882  40.92542  44.3296     5
#  sfh 121.54546 126.58981 143.02528 147.40794 154.30767 165.2755     5

microbenchmark::microbenchmark(
  sf = { sf::st_coordinates( pts ) }
  , sfh = { sfheaders::sf_to_df( pts, fill = TRUE ) }
  , times = 5
)

# Unit: milliseconds
# expr       min        lq     mean    median        uq      max neval
#   sf  27.40953  40.92232  94.5978  41.09299  42.78871 320.7755     5
#  sfh 122.26871 142.14373 140.3767 144.97406 145.53768 146.9593     5
```

**Comment by SymbolixAU** [comment]
And re-doing your benchmarks

```r


n = 1e6
df = data.frame(x = rnorm(n)
                , y = rnorm(n))

pts = sfheaders::sf_point(obj = df)

sf = sf::st_coordinates(pts)
sfh = sfheaders::sf_to_df(pts)

all.equal(unname(as.matrix(sfh[, c("x", "y")])), unname(sf))
# [1] TRUE

bench::mark(check = FALSE,
            sf = {
              sf::st_coordinates(pts)
            }
            , sfh = {
              sfheaders::sf_to_df(pts)
            })

res = bench::press(rows = 10 ^ (1:6),
                   bench::mark(check = FALSE,
                               sf = {
                                 sf::st_coordinates(pts[1:rows,])
                               }
                               , sfh = {
                                 sfheaders::sf_to_df(pts[1:rows,])
                               }))

res
ggplot2::autoplot(res)

 res
# A tibble: 12 x 14
   expression   rows      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory time 
   <bch:expr>  <dbl> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list> <lis>
 1 sf         1.00e1 933.62µs   1.03ms  952.        6.36KB    0       476     0   500.01ms <NULL> <df[,… <bch…
 2 sfh        1.00e1 901.02µs   1.15ms  857.        8.24KB    2.11    406     1   473.66ms <NULL> <df[,… <bch…
 3 sf         1.00e2   3.05ms   3.22ms  305.       56.44KB    0       153     0   500.88ms <NULL> <df[,… <bch…
 4 sfh        1.00e2   3.09ms   3.35ms  291.       60.34KB    0       146     0   501.89ms <NULL> <df[,… <bch…
 5 sf         1.00e3  24.02ms  25.72ms   38.9     474.55KB    2.16     18     1   462.14ms <NULL> <df[,… <bch…
 6 sfh        1.00e3  26.02ms  27.94ms   35.6     488.99KB    0        18     0   505.35ms <NULL> <df[,… <bch…
 7 sf         1.00e4  238.7ms 240.95ms    4.14      4.94MB    0         3     0   725.12ms <NULL> <df[,… <bch…
 8 sfh        1.00e4 244.31ms 245.54ms    4.07      5.05MB    2.04      2     1   491.09ms <NULL> <df[,… <bch…
 9 sf         1.00e5    2.54s    2.54s    0.393     47.3MB    0.393     1     1      2.54s <NULL> <df[,… <bch…
10 sfh        1.00e5     2.6s     2.6s    0.385    48.45MB    0.770     1     2       2.6s <NULL> <df[,… <bch…
11 sf         1.00e6    25.8s    25.8s    0.0388  456.92MB    0.775     1    20      25.8s <NULL> <df[,… <bch…
12 sfh        1.00e6   24.72s   24.72s    0.0405  468.37MB    0.769     1    19     24.72s <NULL> <df[,… <bch…
# … with 1 more variable: gc <list>


```

**Comment by Robinlovelace** [comment]
Seems on par with sf in 'my' benchmark but ~4 times slower on yours. Any ideas why?

**Comment by dcooley** [comment]
I think you only go to 100,000 rows?

**Comment by Robinlovelace** [comment]
Ah yes. Interesting the slowdown between 1e6 and 1e7 is still there.

**Comment by dcooley** [comment]
yeah will be interesting to figure out why that's happening.

**Comment by dcooley** [comment]
Actually, my example is only 1e6 as Well!

**Comment by Robinlovelace** [comment]
microbenchmark vs bench issue?

**Comment by Robinlovelace** [comment]
Thanks @SymbolixAU
